src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -1361,6 +1362,64 @@ public static bool IsFileTypeOrUsesFileTypes(this TypeSymbol type)
             return foundType is not null;
         }
 
+        internal static string? AssociatedFileIdentifier(this NamedTypeSymbol type)
+        {
+            if (type is not SourceMemberContainerTypeSymbol { AssociatedSyntaxTree: SyntaxTree tree })
+            {
+                return null;
+            }
+            var ordinal = type.DeclaringCompilation.GetSyntaxTreeOrdinal(tree);
+
+            var pooledBuilder = PooledStringBuilder.GetInstance();
+            var sb = pooledBuilder.Builder;
+            sb.Append('<');
+            AppendFileName(tree, sb);
+            sb.Append('>');
+            sb.Append((char)GeneratedNameKind.FileType);
+            sb.Append(ordinal);
+            sb.Append("__");
+            return pooledBuilder.ToStringAndFree();
+        }
+
+        internal static string GetDisplayFileName(this SyntaxTree tree)
+        {
+            var pooledBuilder = PooledStringBuilder.GetInstance();
+            var sb = pooledBuilder.Builder;
+            AppendFileName(tree, sb);
+            return pooledBuilder.ToStringAndFree();
+        }
+
+        private static void AppendFileName(SyntaxTree tree, StringBuilder sb)
+        {
+            // note: a "file path" for a syntax tree can be basically any string.
+            // since methods like Path.GetFileNameWithoutExtension throw when invalid path characters are found,
+            // we implement more of a heuristic here.
+            var originalFilePath = tree.FilePath;
+            var fileNameStartIndex = Math.Max(originalFilePath.LastIndexOf('/'), originalFilePath.LastIndexOf('\\')) + 1;
```

Comment:
>Math.Max(originalFilePath.LastIndexOf('/'), originalFilePath.LastIndexOf('\\')) [](http://example.com/codeflow?start=37&length=79)

Perhaps `originalFilePath.LastIndexOfAny(pathChars)` where `static readonly char[] pathChars = new[] { '/', '\\' };`.

Reply:
Moved to using FileNameUtilities.